## KerasTuner On Diabetic Data

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns
from sklearn.model_selection import train_test_split

from sklearn.preprocessing import StandardScaler
import warnings
warnings.filterwarnings('ignore')

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/plotly/datasets/refs/heads/master/diabetes.csv')

In [3]:
df.head(2)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0

In [4]:
df.shape

(768, 9)

#### Handling Missing Values

In [5]:
df.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

In [6]:
df.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [7]:
for column in df.columns[1:6]:
    
    df[column] = np.where(df[column]==0 , df[column].mean() , df[column])

In [8]:
df.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  121.681605      72.254807      26.606479  118.660163   
std       3.369578   30.436016      12.115932       9.631241   93.080358   
min       0.000000   44.000000      24.000000       7.000000   14.000000   
25%       1.000000   99.750000      64.000000      20.536458   79.799479   
50%       3.000000  117.000000      72.000000      23.000000   79.799479   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    32.450805                  0.471876   33.240885    0.348958  
std      6.875374                  0.331329   11.760232    0.476951  
min     18.200000                  0.078000   21.000000    0.000000  
25%     27.500000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

#### Info

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    float64
 2   BloodPressure             768 non-null    float64
 3   SkinThickness             768 non-null    float64
 4   Insulin                   768 non-null    float64
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(6), int64(3)
memory usage: 54.1 KB


### Data Balance

In [10]:
df['Outcome'].value_counts()

0    500
1    268
Name: Outcome, dtype: int64

In [11]:
df['Outcome'].value_counts() / len(df['Outcome']) * 100

0    65.104167
1    34.895833
Name: Outcome, dtype: float64

#### Data is  Balance

#### Data split into independent and dependent variable

In [12]:
x = df.iloc[: , :-1]
y = pd.DataFrame(df.iloc[: , -1])

In [13]:
x.shape

(768, 8)

In [14]:
y.shape

(768, 1)

#### Train test split before feature scaling

In [15]:
x_train, x_test , y_train , y_test = train_test_split(x ,y , test_size = 0.25 , random_state = 16 , stratify = y )

In [16]:
print(x_train.shape, x_test.shape , y_train.shape , y_test.shape)

(576, 8) (192, 8) (576, 1) (192, 1)


## Scaling

In [17]:
scaler = StandardScaler()

In [18]:
x_train_sc = pd.DataFrame(scaler.fit_transform(x_train) , columns = x.columns)
x_test_sc = pd.DataFrame(scaler.transform(x_test) , columns = x.columns) 

In [19]:
print(x_train_sc.shape)
print(x_test_sc.shape)

(576, 8)
(192, 8)


In [20]:
print(x_train_sc.shape , x_test_sc.shape , y_train.shape , y_test.shape)

(576, 8) (192, 8) (576, 1) (192, 1)


In [21]:
x_train_sc.head(2)

Pregnancies   Glucose  BloodPressure  SkinThickness   Insulin       BMI  \
0     0.346355  1.436943       0.287148      -0.625916 -0.414143  1.898677   
1     1.806827 -0.688369       1.078047       1.074377 -0.414143  1.884189   

   DiabetesPedigreeFunction       Age  
0                 -0.400647 -0.514652  
1                  1.894109  0.436029

In [22]:
x_test_sc.head(2)

Pregnancies   Glucose  BloodPressure  SkinThickness   Insulin       BMI  \
0     0.346355  0.456030       0.919867       1.487482 -0.326067  0.348538   
1    -0.529928 -0.328701      -0.345571      -0.474767 -0.261626  0.218152   

   DiabetesPedigreeFunction       Age  
0                 -0.556322  0.176752  
1                 -0.472719 -0.601078

In [23]:
y_train

Outcome
391        1
270        1
42         0
654        0
24         1
..       ...
395        0
542        1
223        0
555        0
643        0

[576 rows x 1 columns]

#### Build KerasTuner (HyperparameterTunning)

In [24]:
import tensorflow as tf
from tensorflow import keras
from keras.layers import *
from keras import Sequential

#### Build Sequential model

In [25]:
def sequential_model(hyp):
    
    model = Sequential()
    
    counter = 0 # This counter for to just check is there input layer
    
    for i in range(hyp.Int('Number_Of_Layers:' , min_value = 1 , max_value = 5)):
        
        
        if counter == 0 :
            
            model.add(Dense(units=hyp.Int('Number_Of_Neurons' + str(i),min_value=4,max_value=64,step=8),
                       activation=hyp.Choice('Activation_Fun:'+str(i), 
                        values = ['relu','leaky_relu','sigmoid','tanh']) ,
                           input_dim = x_train_sc.shape[1]))
            
            
            
            
        else:
            
            model.add(Dense(units=hyp.Int('Number_Of_Neurons'+ str(i),min_value =4,max_value=64,step=8),
                       activation=hyp.Choice('Activation_Fun:'+str(i), 
                                    values = ['relu' , 'leaky_relu' , 'sigmoid' ,'tanh'])))
            
            
#         model.add(Dropout(hyp.Choice('Dropout % is :' , 
#                             values = [0.5,0.10,0.15,0.20,0.25,0.30,0.35,0.40,0.45,0.50])))
            
        
        counter = counter + 1 # increament count of counter
        
        
        
    # Output layer
    model.add(Dense(units = 1 , activation = 'sigmoid'))
    
    model.compile(optimizer = hyp.Choice('Optimizer:+ str(i)' , 
                        values = ['adam' , 'rmsprop' , 'sgd']) ,
                loss = 'binary_crossentropy',
                 metrics = ['accuracy'])
            
            
            
    return model

#### Tune KerasTuner (tunning Sequential model by using kerasTuner)

In [26]:
import keras_tuner as kt

In [27]:
tuner = kt.RandomSearch(hypermodel = sequential_model , 
               objective = 'val_accuracy',
               max_trials = 12,
               directory = 'diabetic_kerasTuner_2',
               project_name = 'DNN_tunning_diabetic_2')

#### If you are facing any issue about value error and show shape is not matched so change
#### directory name and change project name . it will resolve
#max_trials = 10 , means it 10 times build the model and performance evaluation for 10 times and
# gives best accuracy / performance from it

#### kerasTuner is ready for train and search the best model and performs 10 epochs(per trial) and get best accuracy from it.
#### here each trial perform 10 epochs so here it perform 80 epochs .I want to perform 10 epochs for every trial
#### It train kerastuner for 8 times because we give 8 trials means model get build for 8 times and give accuracy for each 8 time and gives final best accuracy on overall performance

In [28]:
#tuner.search(x_train, y_train , epochs = 10 , validation_data = (x_test , y_test))
tuner.search(x_train_sc, y_train, epochs=10, validation_data=(x_test_sc, y_test))

Trial 12 Complete [00h 00m 04s]
val_accuracy: 0.7708333134651184

Best val_accuracy So Far: 0.796875
Total elapsed time: 00h 00m 45s


#### 12 trial accuracy is 0.6510416865348816 and best accuracy from all trials is 0.7708333134651184 (with Dropout)
#### 12 trial accuracy is 0.7708333134651184 and best accuracy from all trials is 0.796875(without Dropout)

In [29]:
print(y_train.shape)
print(y_test.shape)
print(x_train_sc.shape)
print(x_test_sc.shape)
print(x_train_sc.shape[1])

(576, 1)
(192, 1)
(576, 8)
(192, 8)
8


## Find best hyperparameters from this list of best hyperparameters

##### From the list 0 index having best hyperparameters

In [30]:
tuner.get_best_hyperparameters()[0].values

{'Number_Of_Layers:': 2,
 'Number_Of_Neurons0': 36,
 'Activation_Fun:0': 'tanh',
 'Optimizer:+ str(i)': 'rmsprop',
 'Number_Of_Neurons1': 36,
 'Activation_Fun:1': 'leaky_relu',
 'Number_Of_Neurons2': 4,
 'Activation_Fun:2': 'leaky_relu',
 'Number_Of_Neurons3': 52,
 'Activation_Fun:3': 'tanh'}

#### Here , as per  kerastuner , how many layers need to be create i.e. 1 layer . (First lineindicates how many layers suugest by kerastuner) (with Dropout)
#### and rest of lines are nothing but we give 5 layers in for loop so it just execute 5 layers naturally and give best hyperparamter for each layer as per keras (with Dropout)

##### Here , as per kerastuner , how many layers need to be create i.e. 5 layer . (First lineindicates how many layers suugest by kerastuner) (without Dropout)
##### and rest of lines are nothing but we give 5 layers in for loop so it just execute 5 layers naturally and give best hyperparamter for each layer as per keras (with Dropout)¶


## Find the best model among all 8 models 

In [31]:
best_model = tuner.get_best_models(num_models = 1)[0]

In [32]:
best_model

<Sequential name=sequential, built=True>

#### Now our model is ready to fit means ready for training

In [33]:
history = best_model.fit(x_train_sc , y_train , epochs = 50 , initial_epoch = 10 , batch_size = 40 , validation_data = (x_test_sc , y_test))

Epoch 11/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 1s 21ms/step - accuracy: 0.7876 - loss: 0.4689 - val_accuracy: 0.7812 - val_loss: 0.4308
Epoch 12/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7936 - loss: 0.4555 - val_accuracy: 0.7865 - val_loss: 0.4311
Epoch 13/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7925 - loss: 0.4504 - val_accuracy: 0.7812 - val_loss: 0.4334
Epoch 14/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7797 - loss: 0.4628 - val_accuracy: 0.7865 - val_loss: 0.4320
Epoch 15/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.8065 - loss: 0.4439 - val_accuracy: 0.7865 - val_loss: 0.4323
Epoch 16/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7917 - loss: 0.4579 - val_accuracy: 0.7865 - val_loss: 0.4317
Epoch 17/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7923 - loss: 0.4545 - val_accuracy: 0.7812 - val_loss: 0.4297
Epoch 18/50
15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7895 - loss: 0.4459 - val_accuracy: 0.7812 - 

#### We get accuracy of MLP model by kerasTuner (hyperparameters selected by keras) - accuracy: 0.7358 - loss: 0.4918 - val_accuracy: 0.7708 - val_loss: 0.4548 (with Dropout)
####  We get accuracy of MLP model by kerasTuner (hyperparameters selected by keras) -accuracy: 0.7880 - loss: 0.4385 - val_accuracy: 0.7812 - val_loss: 0.4340 (Without Dropout)

### Just For Referance How for loop working in above code where we build hidden layers and output layers and compiled it



Let's break it down step by step based on what you've described:

### 1. **Purpose of the `for` loop**:
   - **Adding Hidden Layers**: 
     - The `for` loop is responsible for adding hidden layers (including their corresponding **Dropout** layers). 
     - You are using `hyp.Int('Number_Of_Layers:', min_value=1, max_value=5)` to define how many hidden layers you want (you are choosing the number of hidden layers between 1 and 5).
     - Inside the loop, you are checking if it's the first layer (`counter == 0`), in which case, you add the **input layer**. For subsequent iterations, you add **hidden layers** (without the `input_dim`).

   - **Hidden Layers**: 
     - You are indeed adding the first hidden layer during the first iteration (`counter == 0`), and for the rest of the iterations (when `counter != 0`), you are adding additional hidden layers.

### 2. **Why the output layer should be outside the loop**:
   - The **output layer** should be added **once**, after all hidden layers are defined. It should **not** be inside the loop because:
     - If you put the output layer inside the loop, it would be added after each hidden layer, which is incorrect. You only want **one output layer** in a neural network, so adding it inside the loop would result in multiple output layers (which is not desired and would raise an error).
   - Therefore, the **output layer** should be added after the loop ends, which is **outside** the loop.

### 3. **Why model compilation should be outside the loop**:
   - **Model Compilation** happens once after the entire architecture is built, which includes both the hidden layers and the output layer. 
   - If you compile the model inside the loop, it would be compiled multiple times during each iteration, which is inefficient and unnecessary.
   - **Only compile the model once**, after the entire architecture is defined.

### Recap of the Correct Logic:
- **Inside the loop**: Add the **hidden layers** and their corresponding **Dropout** layers.
  - For the first iteration (`counter == 0`), add the **input layer**.
  - For the rest, add additional **hidden layers**.
  
- **Outside the loop**: 
  - Add the **output layer** (only once).
  - **Compile the model** (only once, after all layers are defined).

### Final Points:
- **For Loop**: Correctly handles adding hidden layers (including the input layer during the first iteration and the hidden layers in subsequent iterations).
- **Output Layer**: Should be added **once** after all the hidden layers are added, so it goes **outside the loop**.
- **Model Compilation**: Should be done **once**, after the model architecture is fully defined, so it also goes **outside the loop**.